### ISSUE TITLE
Detection of uncoverable failure

### ISSUE THREAD

**Issue by fortime** [Parsely/pykafka#227]
Will pykafka provide a way to detect uncoverable failure, except the running flag. I want to know if the consumer has ocurred uncoverable failure when I catch an exception. Or, if all exception is coverable.

**Comment by yungchin** [comment]
Hi @fortime, yes, I would think so - imho, any unrecoverable failure that goes undetected should be classified as a bug. I suppose this question ties in with the work you've been doing in #206 - ie stopping the consumer when unexpected exceptions kill a worker thread. Do you have second thoughts about how that works?

**Comment by fortime** [comment]
@yungchin I think it should provide a parent exception which means a uncoverable failure. So that the user of pykafka can determine if he should stop the consumer and start a new one.

BTW, I have found three problems:
1. it should remove all watches and all ephemeral nodes from zookeeper if the zookeeper client is from outside and the consumer is not running.
2. if the autostart flag is set, it should catch all exceptions from `start()`, then stop the consumer. Otherwise, those started threads will run forever.
3. like #223, it should be a broker connection failover problems. It hangs when the producer is producing messages, and [retry logic of producer when socket disconnect exception is catched](https://github.com/Parsely/pykafka/blob/master/pykafka/producer.py#L168-L177) never makes sense.

**Comment by yungchin** [comment]
@fortime many thanks for all the careful analysis you've contributed in this and other tickets and pullreqs that I've just updated. I've been going around splitting them up into separate issues, to make them more tractable, and I'll do the same here, hope that's ok.

I'll be closing this, seeing that the main issue here has several duplicates on the tracker, and work to address it is ongoing at #235. As for the other issues you mention:

  1. I will spin that one out as a new ticket
  2. I suspect this has been addressed by the various thread/connection-leak fixes that have made it to master in the last two weeks
  3. IIUC that's fully covered by #223?

Thanks.
